docs/specs/pipelines-io.md

```diff
@@ -0,0 +1,196 @@
+## Efficient IO
+
+As part of the high performance work, we need new abstractions for doing efficient reading and writing to and from the network. The pattern we have settled on
+so far in pipelines is a model where buffers are pushed by a producer (`IPipelineWriter`) to a consumer (`IPipelineReader`).
+
+```C#
+public interface IPipelineReader
+{
+    ReadableBufferAwaitable ReadAsync();
+    void Advance(ReadCursor consumed, ReadCursor examined);
+    void CancelPendingRead();
+    void Complete(Exception exception = null);
+}
+```
+
+```C#
+public interface IPipelineWriter
+{
+    Task Writing { get; }
+    WritableBuffer Alloc(int minimumSize = 0);
+    void Complete(Exception exception = null);
+}
+```
+
+The producer gets a handle on a `WritableBuffer` and the consumer gets a handle on a `ReadableBuffer` to do writes and reads respectively. 
+
+Memory in pipelines is managed as a linked list of `OwnedMemory<T>`:
+
+```C#
+public class BufferSegment
+{
+    // The backing memory, usually leased from a pool
+    OwnedMemory<byte> Data { get; set; }
+    public int Start { get; set; }
+    public int End { get; set; }
+    
+    // The next segment
+    BufferSegment Next { get; set; }
+    
+    // The segment cannot be written to
+    public bool ReadOnly { get; set; }
+    
+    public int ReadableBytes => End - Start;
+    public int WritableBytes => Data.Length - End;
+}
+```
+
+A `BufferSegment` represents a link in a contiguous chain of buffers. New segments are added if the requested memory is less than the writable bytes in the current segment. What's interesting here is that there are 2 views to a `BufferSegment`, the ReadableBytes and the WritableBytes. The writing and reading primitives in pipelines build on this concept.
+
+### Writing
+
+```C#
+private Task Produce(IPipelineWriter writer)
+{
+   while (true)
+   {
+       // Allocate a buffer from the writer
+       WritableBuffer buffer = writer.Alloc();
+       
+       var bytes = Encoding.UTF8.GetBytes(DateTime.Now.ToString());
+       buffer.Write(bytes);
+
+       // Tell the producer we have data
+       await buffer.FlushAsync();
+       
+       // Wait a second before writing again
+       await Task.Delay(1000);
+   }
+}
+```
+
+In the above example, `Produce` calls `Alloc` on the `IPipelineWriter` to get access to the `WritableBuffer`, then it calls `FlushAsync` to signal to the consumer that data is ready. A `WritableBuffer` is a struct that allows writing to an *ever* growing sink of bytes.
+
+```C#
+public struct WritableBuffer
+{
+   public Memory<byte> Memory { get; }
+   public void Ensure(int minimumBytes = 0);
+   public void Advance(int bytesWritten);
+   public void Commit();
+   public Task FlushAsync();
+}
+```
+
+There are extension methods like `Write(ReadOnlySpan<byte>)` which automate writing to the available memory and advancing the buffer.
+
+```C#
+public static void Write(this WritableBuffer buffer, ReadOnlySpan<byte> source)
```

Comment:
I think this method might become the pit of failure. People should be writing directly to WritableBuffer, not first to span and then copy to the buffer. I know there are cases where the data written is static and just needs to be copied, but: a) we should first show samples how to write directly to buffer, b) possibly rename this method to Copy, as it's meant only for copy scenarios.